src/frontend/src/optimizer/rule/top_n_on_index_rule.rs

```diff
@@ -0,0 +1,77 @@
+//  Copyright 2023 RisingWave Labs
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+// This source code is licensed under both the GPLv2 (found in the
+// COPYING file in the root directory) and Apache 2.0 License
+// (found in the LICENSE.Apache file in the root directory).
+use crate::optimizer::plan_node::{LogicalLimit, LogicalScan, LogicalTopN, PlanTreeNodeUnary};
+use crate::optimizer::property::Direction::Asc;
+use crate::optimizer::property::{FieldOrder, Order};
+use crate::optimizer::PlanRef;
+
+pub struct TopNOnIndexRule {}
+
+impl Rule for TopNOnIndexRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let logical_topn: &LogicalTopN = plan.as_logical_top_n()?;
+        let logical_scan: LogicalScan = logical_topn.input().as_logical_scan()?.to_owned();
+        let order = logical_topn.topn_order();
+        if order.field_order.is_empty() {
+            return None;
+        }
+        let index = logical_scan.indexes().iter().find(|idx| {
+            Order {
+                field_order: idx
+                    .index_item
+                    .iter()
+                    .map(|idx_item| FieldOrder {
+                        index: idx_item.index,
+                        direct: Asc,
+                    })
```

Comment:
I'm not sure whether we've supported it, but we can specify `DESC` order when creating index in postgres (which is often used along with `NULLS FIRST/LAST`). So maybe we should not hard-code an `Asc` here, but use `index_table.pk` instead.

cc @chenzl25

Reply:
Agree. Use `index_table.pk` instead of hard-code `Asc` even if we do not support creating indexes with descending ordering currently.

Reply:
Since `index_items` and `index_table.pk.items()` are not quite matched, I'm using `index_table.pk.items()` instead and it seems won't affect the result. Actually this part of code is derived from https://github.com/risingwavelabs/risingwave/blob/ed27ecdbdcd835f4db00da6ff575377006b24c86/src/frontend/src/optimizer/plan_node/logical_scan.rs#L551. Shall I also change the logic there?